packages/i18n-calypso/types/index.d.ts

```diff
@@ -56,7 +60,7 @@ export type TranslateOptionsPluralText = TranslateOptionsPlural & { textOnly: tr
 
 // Translate hooks, like component interpolation or highlighting untranslated strings,
 // force us to declare the return type as a generic React node, not as just string.
-export type TranslateResult = React.ReactChild;
```

Comment:
The ReactChild types are deprecated, and they also don't seem to work any more. Since the intention originally was to avoid using ReactNode because of its falsey type options, I changed to use ExistingReactNode which filters the types on ReactNode.

Reply:
One thing that isn't clear to me is what we actually expect the type to be. 

Here are the React 18 type definitions (updated since React 17!):

```ts
type ReactNode = ReactElement | string | number | ReactFragment | ReactPortal | boolean | null | undefined

type ReactFragment = Iterable<ReactNode`
interface ReactPortal extends ReactElement {
  key: Key | null;
  children: ReactNode;
}
```

`ReactNode` is supposed to represent anything that React can render. (Though I suppose that means anything you can put in JSX, not anything you can return from render!) In this case, I filtered out the null, undefined, and boolean types to create the (admittedly oddly named) `ExistingReactNode` type, which leaves us with `ReactElement | ReactFragment | string | number` (considering that ReactPortal is just a more specific type of ReactElement.` 

But since the result of a translate operation is a variable you would put into jsx (e.g. `<p>{translate( 'hello world ')}</p>`), it seems like any truthy `ReactNode` result would work fine.

Since the current state of the PR is accepting `ReactElement | string | number | ReactFragment | ReactPortal`, I don't know if further filtering the type to just `ReactElement | string` would resolve any of the type issues.

Most of the issues I've been running into are that the input args often don't strictly exist! Take this example:

```tsx
{ translate( 'Establishing a connection to {{strong}}%(targetSite)s{{/strong}}.', {
  args: {
    targetSite: targetSite ? targetSite : siteSlug,
  },
  components: {
    strong: <strong />,
  },
} ) }
```

The type issue here is that `siteSlug` is nullable. (Because `useSelector( getSelectedSiteId )` can return null.) Though I suspect that while under this code path, a site is probably selected, the types aren't strong enough to see that.

Similar issues account for many of the types I've been working through.

Reply:
I guess the core question is whether there are any limitations on what substitutions can handle (beyond non-nullable etc). If it can handle inputs that are ReactPortal or ReactFragment, it seems fine to accept those

The core reason we're getting errors now is that the React 17 typed `ReactFragment` as `{} | Iterable<ReactNode>`, which is a super loose type. That's been fixed now, which probably actually makes our types work more correctly now